qualcommax: ipq807x: use ath11k_patch_mac and ath11k_remove_regdomain for MX4200 and RAX120v2

testuser7 opened the pull request:
Add new functions for ath11k caldata:
- ath11k_patch_mac (from 0 to 5)
- ath11k_remove_regdomain
- ath11k_set_macflag (some pre-caldata have the nvMacFlag flag unset which is needed to change the MAC address)

and use them for:
- MX4200 (only v2 variant requires MAC patching)
- RAX120v2 (pre-caldata does not contain valid MAC addresses)

Additionally for ath10k caldata:
- ath10k_remove_regdomain

testuser7 committed:
base-files: Add new functions for ath11k caldata

Add new functions for ath11k caldata:
- ath11k_patch_mac (from 0 to 5)
- ath11k_remove_regdomain
- ath11k_set_macflag (some pre-caldata have the nvMacFlag flag unset which is needed to change the MAC address)

Additionaly for ath10k caldata:
- ath10k_remove_regdomain

Signed-off-by: Paweł Owoc <frut3k7@gmail.com>

testuser7 committed:
qualcommax: ipq807x: use ath11k_patch_mac and ath11k_remove_regdomain for MX4200

Use ath11k_patch_mac, ath11k_remove_regdomain and ath11k_set_macflag functions for MX4200
(only v2 variant requires MAC patching)

Signed-off-by: Paweł Owoc <frut3k7@gmail.com>

testuser7 commented:
This has been verified for 3 routers:

1. MX5300 - removing regdomian only from the `caldata` is sufficient (regdomain is set in 4 addresses in `pre-caldata` and `caldata`)
2. MX4200v1 - after removing regdomain only from `caldata` we are able to change country but tx-power is very low (regdomain is set in 4 addresses in `pre-caldata` and `caldata`)
3. MX4200v2 - same as MX4200v1 (regdomain is set only in 1 address in `pre-caldata` and `caldata`)

Ansuel commented:
@testuser7 can you by chance share with me the MX4200v1 pre-cal and cal? I'm curious if i manage to understand the logic the fw apply to enforce the regdomain.

testuser7 commented:
> @testuser7 can you by chance share with me the MX4200v1 pre-cal and cal? I'm curious if i manage to understand the logic the fw apply to enforce the regdomain.

I don't have this device.
MX4200v1 `caldata`: [bdwlan.bin.gz](https://github.com/openwrt/openwrt/files/14100648/bdwlan.bin.gz)
@SpectreDev1 Can you help with `pre-caldata`?

SpectreDev1 commented:
> > @testuser7 can you by chance share with me the MX4200v1 pre-cal and cal? I'm curious if i manage to understand the logic the fw apply to enforce the regdomain.
> 
> I don't have this device. MX4200v1 `caldata`: [bdwlan.bin.gz](https://github.com/openwrt/openwrt/files/14100648/bdwlan.bin.gz) @SpectreDev1 Can you help with `pre-caldata`?

If I can find some free time I'll help, just quite busy currently

Razerwire commented:
@testuser7 Do these patches this one and [#27](https://github.com/openwrt/firmware_qca-wireless/pull/27) ) (still) limit 5Ghz TX-Power  to dBm?

testuser7 commented:
> @testuser7 Do these patches this one and [#27](https://github.com/openwrt/firmware_qca-wireless/pull/27) ) (still) limit 5Ghz TX-Power to 8 dBm?

No, it is no longer limited but there are some limits for ETSI region. See here: https://github.com/openwrt/firmware_qca-wireless/pull/23

testuser7 commented:
> > @testuser7 can you by chance share with me the MX4200v1 pre-cal and cal? I'm curious if i manage to understand the logic the fw apply to enforce the regdomain.
> 
> @Ansuel At the request of @testuser7; i'm happy to share a copy of my pre-cal data (from a MX4200v1 EU)
> 
> The pre-cal-data resides in `"mtd17: 00080000 00020000 "0:art"` , am i correct?

`caldata_extract "0:art" 0x1000 0x20000`

Razerwire commented:
> caldata_extract "0:art" 0x1000 0x20000
@testuser7 @Ansuel 
[cal-ahb-c000000.wifi.bin.tar.gz](https://github.com/openwrt/openwrt/files/14143614/cal-ahb-c000000.wifi.bin.tar.gz)

This should be it (see attached tar.gz)  please do let me know if i have failed)

testuser7 commented:
> > [cal-ahb-c000000.wifi.bin.tar.gz](https://github.com/openwrt/openwrt/files/14143614/cal-ahb-c000000.wifi.bin.tar.gz)
> 
> This should be it (see attached tar.gz) please do let me know if i have failed in providing the proper data)

This `pre-caldata` is correct.
@Ansuel did you check the fw logic?

robimarko commented:
@Ansuel What do you think?

I personally find the MAC patch quite interesting as it would allow us to drop the current hotplug script

testuser7 commented:
@Ansuel Any comments on this PR from you?

testuser7 committed:
qualcommax: ipq807x: use ath11k_patch_mac for RAX120v2

Use ath11k_patch_mac and ath11k_set_macflag functions for RAX120v2 (pre-caldata does not contain valid MAC addresses)

Signed-off-by: Paweł Owoc <frut3k7@gmail.com>

ilkka-ollakka commented:
For me in MX4200v1 without this PR I only got ~5-8dBm of tx in any of the radios in FI country-code, with this applied, I was able to get normal tx powers in all radios.

robimarko commented:
The regulatory domain was removed from the MX4200 BDF in:
https://github.com/openwrt/firmware_qca-wireless/commit/fc30aeeb8ec5d069710a446f4eb5e30fc26145c1

testuser7 commented:
> The regulatory domain was removed from the MX4200 BDF in: [openwrt/firmware_qca-wireless@fc30aee](https://github.com/openwrt/firmware_qca-wireless/commit/fc30aeeb8ec5d069710a446f4eb5e30fc26145c1)

We need to remove regdomain from `pre-caldata` to make it work.

robimarko commented:
Has this refactor been tested on ath10k as well?

testuser7 commented:
> Has this refactor been tested on ath10k as well?

For `ath10k` with `GB` regdomain in pre-caldata:

1. Without patch (country `US`):

```
root@OpenWrt:~# iw reg get
global
country US: DFS-FCC
	(902 - 904 @ 2), (N/A, 30), (N/A)
	(904 - 920 @ 16), (N/A, 30), (N/A)
	(920 - 928 @ 8), (N/A, 30), (N/A)
	(2400 - 2472 @ 40), (N/A, 30), (N/A)
	(5150 - 5250 @ 80), (N/A, 23), (N/A), AUTO-BW
	(5250 - 5350 @ 80), (N/A, 24), (0 ms), DFS, AUTO-BW
	(5470 - 5730 @ 160), (N/A, 24), (0 ms), DFS
	(5730 - 5850 @ 80), (N/A, 30), (N/A), AUTO-BW
	(5850 - 5895 @ 40), (N/A, 27), (N/A), NO-OUTDOOR, AUTO-BW, PASSIVE-SCAN
	(5925 - 7125 @ 320), (N/A, 12), (N/A), NO-OUTDOOR, PASSIVE-SCAN
	(57240 - 71000 @ 2160), (N/A, 40), (N/A)

phy#0
country GB: DFS-ETSI
	(2400 - 2483 @ 40), (N/A, 20), (N/A)
	(5150 - 5250 @ 80), (N/A, 23), (N/A), NO-OUTDOOR, AUTO-BW
	(5250 - 5350 @ 80), (N/A, 20), (0 ms), NO-OUTDOOR, DFS, AUTO-BW
	(5470 - 5730 @ 160), (N/A, 26), (0 ms), DFS
	(5725 - 5850 @ 80), (N/A, 23), (N/A), NO-OUTDOOR
	(5925 - 6425 @ 160), (N/A, 23), (N/A), NO-OUTDOOR
	(57000 - 71000 @ 2160), (N/A, 40), (N/A)
```
```
Thu Mar 14 12:49:30 2024 daemon.notice netifd: radio0 (2878): WARNING: Variable 'data' does not exist or is not an array/object
Thu Mar 14 12:49:30 2024 daemon.notice hostapd: Set new config for phy phy0:
Thu Mar 14 12:49:30 2024 daemon.notice wpa_supplicant[1855]: Set new config for phy phy0
Thu Mar 14 12:49:30 2024 daemon.notice wpa_supplicant[1855]: Set new config for phy phy0
Thu Mar 14 12:49:30 2024 daemon.notice hostapd: Set new config for phy phy0: /var/run/hostapd-phy0.conf
Thu Mar 14 12:49:30 2024 daemon.notice hostapd: Restart interface for phy phy0
Thu Mar 14 12:49:30 2024 daemon.notice hostapd: Configuration file: data: driver=nl80211 logger_syslog=127 logger_syslog_level=2 logger_stdout=127 logger_stdout_level=2 country_code=US ieee80211d=1 ieee80211h=1 hw_mode=a beacon_int=100 chanlist=149 tx_queue_data2_burst=2.0 #num_global_macaddr=1 ieee80211n=1 ht_coex=0 ht_capab=[HT40+][LDPC][SHORT-GI-20][SHORT-GI-40][TX-STBC][RX-STBC1][MAX-AMSDU-7935][DSSS_CCK-40] ieee80211ac=1 vht_oper_chwidth=1 vht_oper_centr_freq_seg0_idx=155 vht_capab=[RXLDPC][SHORT-GI-80][TX-STBC-2BY1][SU-BEAMFORMER][SU-BEAMFORMEE][MU-BEAMFORMER][MU-BEAMFORMEE][RX-ANTENNA-PATTERN][TX-ANTENNA-PATTERN][RX-STBC-1][SOUNDING-DIMENSION-4][BF-ANTENNA-4][MAX-MPDU-11454][MAX-A-MPDU-LEN-EXP7] channel=149  interface=phy0-ap0 bssid=e8:9f:80:5f:d4:1b ctrl_interface=/var/run/hostapd ap_isolate=1 bss_load_update_period=60 chan_util_avg_period=600 disassoc_low_ack=1 skip_inactivity_poll=0 preamble=1 wmm_enabled=1 ignore_broadcast_ssid=0 uapsd_advertisement_enabled=1 utf8_ssid=1 multi_ap=0 sae_require_mfp=1 sae_pwe=2 wpa_pa
Thu Mar 14 12:49:36 2024 kern.warn kernel: [   92.179288] ath10k_pci 0001:01:00.0: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
Thu Mar 14 12:49:36 2024 kern.warn kernel: [   92.179354] ath10k_pci 0001:01:00.0: msdu-desc: 2500  skid: 32
Thu Mar 14 12:49:36 2024 kern.info kernel: [   92.262444] ath10k_pci 0001:01:00.0: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
Thu Mar 14 12:49:36 2024 kern.info kernel: [   92.263255] ath10k_pci 0001:01:00.0: wmi print 'free: 84920 iram: 13156 sram: 11224'
Thu Mar 14 12:49:36 2024 kern.info kernel: [   92.647314] ath10k_pci 0001:01:00.0: rts threshold -1
Thu Mar 14 12:49:36 2024 kern.warn kernel: [   92.647646] ath10k_pci 0001:01:00.0: Firmware lacks feature flag indicating a retry limit of > 2 is OK, requested limit: 4
Thu Mar 14 12:49:36 2024 kern.info kernel: [   92.652684] br-lan: port 5(phy0-ap0) entered blocking state
Thu Mar 14 12:49:36 2024 kern.info kernel: [   92.662320] br-lan: port 5(phy0-ap0) entered disabled state
Thu Mar 14 12:49:36 2024 kern.info kernel: [   92.667996] device phy0-ap0 entered promiscuous mode
Thu Mar 14 12:49:36 2024 kern.info kernel: [   92.673397] br-lan: port 5(phy0-ap0) entered blocking state
Thu Mar 14 12:49:36 2024 kern.info kernel: [   92.678542] br-lan: port 5(phy0-ap0) entered forwarding state
Thu Mar 14 12:49:36 2024 kern.info kernel: [   92.683975] br-lan: port 5(phy0-ap0) entered disabled state
Thu Mar 14 12:49:36 2024 daemon.notice hostapd: phy0-ap0: interface state UNINITIALIZED->COUNTRY_UPDATE
Thu Mar 14 12:49:36 2024 daemon.notice hostapd: phy0-ap0: interface state COUNTRY_UPDATE->HT_SCAN
Thu Mar 14 12:49:36 2024 daemon.notice netifd: Wireless device 'radio0' is now up
Thu Mar 14 12:49:36 2024 kern.info kernel: [   92.856817] ath10k_pci 0001:01:00.0: NOTE:  Firmware DBGLOG output disabled in debug_mask: 0x10000000
Thu Mar 14 12:49:37 2024 kern.info kernel: [   93.348403] IPv6: ADDRCONF(NETDEV_CHANGE): phy0-ap0: link becomes ready
Thu Mar 14 12:49:37 2024 kern.info kernel: [   93.348587] br-lan: port 5(phy0-ap0) entered blocking state
Thu Mar 14 12:49:37 2024 kern.info kernel: [   93.353833] br-lan: port 5(phy0-ap0) entered forwarding state
Thu Mar 14 12:49:37 2024 daemon.notice netifd: Network device 'phy0-ap0' link is up
Thu Mar 14 12:49:37 2024 daemon.notice hostapd: phy0-ap0: interface state HT_SCAN->ENABLED
Thu Mar 14 12:49:37 2024 daemon.notice hostapd: phy0-ap0: AP-ENABLED
```

2. and with patch (country `US`):

```
root@OpenWrt:~# iw reg get
global
country US: DFS-FCC
	(902 - 904 @ 2), (N/A, 30), (N/A)
	(904 - 920 @ 16), (N/A, 30), (N/A)
	(920 - 928 @ 8), (N/A, 30), (N/A)
	(2400 - 2472 @ 40), (N/A, 30), (N/A)
	(5150 - 5250 @ 80), (N/A, 23), (N/A), AUTO-BW
	(5250 - 5350 @ 80), (N/A, 24), (0 ms), DFS, AUTO-BW
	(5470 - 5730 @ 160), (N/A, 24), (0 ms), DFS
	(5730 - 5850 @ 80), (N/A, 30), (N/A), AUTO-BW
	(5850 - 5895 @ 40), (N/A, 27), (N/A), NO-OUTDOOR, AUTO-BW, PASSIVE-SCAN
	(5925 - 7125 @ 320), (N/A, 12), (N/A), NO-OUTDOOR, PASSIVE-SCAN
	(57240 - 71000 @ 2160), (N/A, 40), (N/A)

phy#0
country US: DFS-FCC
	(902 - 904 @ 2), (N/A, 30), (N/A)
	(904 - 920 @ 16), (N/A, 30), (N/A)
	(920 - 928 @ 8), (N/A, 30), (N/A)
	(2400 - 2472 @ 40), (N/A, 30), (N/A)
	(5150 - 5250 @ 80), (N/A, 23), (N/A), AUTO-BW
	(5250 - 5350 @ 80), (N/A, 24), (0 ms), DFS, AUTO-BW
	(5470 - 5730 @ 160), (N/A, 24), (0 ms), DFS
	(5730 - 5850 @ 80), (N/A, 30), (N/A), AUTO-BW
	(5850 - 5895 @ 40), (N/A, 27), (N/A), NO-OUTDOOR, AUTO-BW, PASSIVE-SCAN
	(5925 - 7125 @ 320), (N/A, 12), (N/A), NO-OUTDOOR, PASSIVE-SCAN
	(57240 - 71000 @ 2160), (N/A, 40), (N/A)
```
```
Thu Mar 14 12:43:02 2024 daemon.notice netifd: radio0 (2845): WARNING: Variable 'data' does not exist or is not an array/object
Thu Mar 14 12:43:02 2024 daemon.notice hostapd: Set new config for phy phy0:
Thu Mar 14 12:43:02 2024 daemon.notice wpa_supplicant[1822]: Set new config for phy phy0
Thu Mar 14 12:43:02 2024 daemon.notice wpa_supplicant[1822]: Set new config for phy phy0
Thu Mar 14 12:43:02 2024 daemon.notice hostapd: Set new config for phy phy0: /var/run/hostapd-phy0.conf
Thu Mar 14 12:43:02 2024 daemon.notice hostapd: Restart interface for phy phy0
Thu Mar 14 12:43:02 2024 daemon.notice hostapd: Configuration file: data: driver=nl80211 logger_syslog=127 logger_syslog_level=2 logger_stdout=127 logger_stdout_level=2 country_code=US ieee80211d=1 ieee80211h=1 hw_mode=a beacon_int=100 chanlist=149 tx_queue_data2_burst=2.0 #num_global_macaddr=1 ieee80211n=1 ht_coex=0 ht_capab=[HT40+][LDPC][SHORT-GI-20][SHORT-GI-40][TX-STBC][RX-STBC1][MAX-AMSDU-7935][DSSS_CCK-40] ieee80211ac=1 vht_oper_chwidth=1 vht_oper_centr_freq_seg0_idx=155 vht_capab=[RXLDPC][SHORT-GI-80][TX-STBC-2BY1][SU-BEAMFORMER][SU-BEAMFORMEE][MU-BEAMFORMER][MU-BEAMFORMEE][RX-ANTENNA-PATTERN][TX-ANTENNA-PATTERN][RX-STBC-1][SOUNDING-DIMENSION-4][BF-ANTENNA-4][MAX-MPDU-11454][MAX-A-MPDU-LEN-EXP7] channel=149  interface=phy0-ap0 bssid=e8:9f:80:5f:d4:1b ctrl_interface=/var/run/hostapd ap_isolate=1 bss_load_update_period=60 chan_util_avg_period=600 disassoc_low_ack=1 skip_inactivity_poll=0 preamble=1 wmm_enabled=1 ignore_broadcast_ssid=0 uapsd_advertisement_enabled=1 utf8_ssid=1 multi_ap=0 sae_require_mfp=1 sae_pwe=2 wpa_pa
Thu Mar 14 12:43:08 2024 kern.warn kernel: [   73.368486] ath10k_pci 0001:01:00.0: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
Thu Mar 14 12:43:08 2024 kern.warn kernel: [   73.368537] ath10k_pci 0001:01:00.0: msdu-desc: 2500  skid: 32
Thu Mar 14 12:43:08 2024 kern.info kernel: [   73.451570] ath10k_pci 0001:01:00.0: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
Thu Mar 14 12:43:08 2024 kern.info kernel: [   73.452373] ath10k_pci 0001:01:00.0: wmi print 'free: 84920 iram: 13156 sram: 11224'
Thu Mar 14 12:43:08 2024 kern.info kernel: [   73.837600] ath10k_pci 0001:01:00.0: rts threshold -1
Thu Mar 14 12:43:08 2024 kern.warn kernel: [   73.837928] ath10k_pci 0001:01:00.0: Firmware lacks feature flag indicating a retry limit of > 2 is OK, requested limit: 4
Thu Mar 14 12:43:08 2024 kern.info kernel: [   73.842947] br-lan: port 5(phy0-ap0) entered blocking state
Thu Mar 14 12:43:08 2024 kern.info kernel: [   73.852597] br-lan: port 5(phy0-ap0) entered disabled state
Thu Mar 14 12:43:08 2024 kern.info kernel: [   73.858276] device phy0-ap0 entered promiscuous mode
Thu Mar 14 12:43:08 2024 kern.info kernel: [   73.863674] br-lan: port 5(phy0-ap0) entered blocking state
Thu Mar 14 12:43:08 2024 kern.info kernel: [   73.868815] br-lan: port 5(phy0-ap0) entered forwarding state
Thu Mar 14 12:43:08 2024 kern.info kernel: [   73.874273] br-lan: port 5(phy0-ap0) entered disabled state
Thu Mar 14 12:43:08 2024 daemon.notice hostapd: phy0-ap0: interface state UNINITIALIZED->COUNTRY_UPDATE
Thu Mar 14 12:43:08 2024 daemon.notice hostapd: phy0-ap0: interface state COUNTRY_UPDATE->HT_SCAN
Thu Mar 14 12:43:08 2024 kern.debug kernel: [   73.880741] ath: EEPROM regdomain: 0x8348
Thu Mar 14 12:43:08 2024 kern.debug kernel: [   73.880758] ath: EEPROM indicates we should expect a country code
Thu Mar 14 12:43:08 2024 kern.debug kernel: [   73.880763] ath: doing EEPROM country->regdmn map search
Thu Mar 14 12:43:08 2024 kern.debug kernel: [   73.880766] ath: country maps to regdmn code: 0x3a
Thu Mar 14 12:43:08 2024 kern.debug kernel: [   73.880769] ath: Country alpha2 being used: US
Thu Mar 14 12:43:08 2024 kern.debug kernel: [   73.880773] ath: Regpair used: 0x3a
Thu Mar 14 12:43:08 2024 kern.debug kernel: [   73.880776] ath: regdomain 0x8348 dynamically updated by user
Thu Mar 14 12:43:08 2024 daemon.notice netifd: Wireless device 'radio0' is now up
Thu Mar 14 12:43:09 2024 kern.info kernel: [   74.048160] ath10k_pci 0001:01:00.0: NOTE:  Firmware DBGLOG output disabled in debug_mask: 0x10000000
Thu Mar 14 12:43:09 2024 kern.info kernel: [   74.539204] IPv6: ADDRCONF(NETDEV_CHANGE): phy0-ap0: link becomes ready
Thu Mar 14 12:43:09 2024 kern.info kernel: [   74.539382] br-lan: port 5(phy0-ap0) entered blocking state
Thu Mar 14 12:43:09 2024 kern.info kernel: [   74.544662] br-lan: port 5(phy0-ap0) entered forwarding state
Thu Mar 14 12:43:09 2024 daemon.notice netifd: Network device 'phy0-ap0' link is up
Thu Mar 14 12:43:09 2024 daemon.notice hostapd: phy0-ap0: interface state HT_SCAN->ENABLED
Thu Mar 14 12:43:09 2024 daemon.notice hostapd: phy0-ap0: AP-ENABLED
```

But I don't know if this change affects the operation of the radio.

testuser7 commented:
@Ansuel, did you manage to check the logic?
There are more devices where you need to set MAC addresses.

ricsc commented on package/base-files/files/lib/functions/caldata.sh:
```diff
@@ -70,7 +70,7 @@ caldata_extract_reverse() {
 
 	mtd=$(find_mtd_chardev "$part")
-	reversed=$(hexdump -v -s $offset -n $count -e '/1 "%02x "' $mtd)
+	reversed=$(hexdump -v -s $offset -n $count -e '1/1 "%02x "' $mtd)
```
This change breaks caldata at least on zyxel,nbg7815 I tested this without using any of the auxiliary functions like caldata_patch etc. purely using caldata_extract

testuser7 replied:
What is the output for this two commands?
```
hexdump -v -s 0x1000 -n 10 -e '1/1 "%02x "' /dev/mtd17
hexdump -v -s 0x1000 -n 10 -e '/1 "%02x "' /dev/mtd17
```

pwned-pixel replied:
I can confirm that it breaks sth. for nbg7815. I've tried it yesterday with 6.6 (vanilla, just pulling your 3 commits). I didn't look into it further. So I cannot confirm its this specific line. The output for both commands on my device is identical:

`01 00 04 04 00 00 00 00 00 80`

I just observed that within dmesg there was a line complaining "firmware not found". I'll try to make a new build in a few hours and try to collect more information if possible.

testuser7 replied:
Run this bash script and compare checksum:
```
#!/bin/sh

caldata=""

reversed=$(hexdump -v -s 0x1000 -n 131072 -e '1/1 "%02x "' /dev/mtd17)

for byte in $reversed; do
	caldata="\x${byte}${caldata}"
done
printf "%b" "$caldata" > /tmp/pre-caldata.bin

caldata=""

reversed=$(hexdump -v -s 0x1000 -n 131072 -e '/1 "%02x "' /dev/mtd17)

for byte in $reversed; do
	caldata="\x${byte}${caldata}"
done
printf "%b" "$caldata" > /tmp/pre-caldata_2.bin
```

pwned-pixel replied:
Sth. is wrong in this for-loop. It get stuck forever. $reversed is filled.

ricsc replied:
I now added the 1 again and this time there was no issue, not sure what's going on. But caldata extraction on the armor g5 via hotplug when blockd is installed behaves erroneous at least.

testuser7 replied:
As a correction, this function is not used for ipq807x (it is only used for a few fritz devices).
The function is correct, but in the case of large amounts of data, the execution time is very long (in this case, over 4 minutes).
A change in the code does not affect the data.

pwned-pixel replied:
Sorry. I didn't have time to compile yesterday. Today I'm facing issues with gettext. I hope I can figure out the issue.

If I understand you right this is a runtime issue. I didn't wait 4 minutes for this script to finish.

But in general it would make sense and this is not related to your patch. We have already a runtime issue regarding firmware extract with blockd involved for this device (which I didn't use for my test). I will try to make a build as soon as possible and report back if I can reproduce anything. But I don't think its related to your patch after ricsc has indicated it is working. 

The first command is faster (like 2 seconds). The second takes like 8-10 seconds. So I would stick to the first.
In my experience cat is a bit faster but I don't know if cat is allowed in those POSIX/sh scripts.

testuser7 replied:
I had a issue once that `pre-caldata` was not extracted running initramfs image on other device. But didn't check what was the issue.

pwned-pixel replied:
Hmm. I never thought about that this could be an issue regarding testing. Anyway. I've made a clean build (tools/toolchain new) with your three commits only. Without any other modification and without luci now. I've loaded it three times through initramfs. Everything got loaded.
After that test I've flashed it to the device. Several reboots. No issue. So I guess sth. fkd things up last time or it was indeed initramfs related.

A side question: How do this fix should look for this device? I've tried:

`zyxel,nbg7815)
	caldata_extract "0:art" 0x1000 0x20000
	label_mac=$(get_mac_label)
	ath11k_patch_mac $(macaddr_add $label_mac 3) 1
	ath11k_patch_mac $(macaddr_add $label_mac 2) 2
	ath11k_patch_mac $(macaddr_add $label_mac 4) 3`

But that did nothing in regard fixing the MAC. But the device came up working with its usual random MAC's anyway.

testuser7 replied:
You must additionally use the `ath11k_fix_caldata` function. `pre-caldata` does not contain real MAC addresses for this device.

testuser7 replied:
@pwned-pixel Can you check if these commands give you correct MAC addresses?
```
hexdump -s 14 -n 6 -C /lib/firmware/ath11k/IPQ8074/hw2.0/cal-ahb-c000000.wifi.bin
hexdump -s 20 -n 6 -C /lib/firmware/ath11k/IPQ8074/hw2.0/cal-ahb-c000000.wifi.bin
hexdump -s 26 -n 6 -C /lib/firmware/ath11k/IPQ8074/hw2.0/cal-ahb-c000000.wifi.bin
```

pwned-pixel replied:
```
root@NBG7815:~# hexdump -s 14 -n 6 -C /lib/firmware/ath11k/IPQ8074/hw2.0/cal-ahb-c000000.wifi.bin
0000000e  00 03 7f 12 34 56                                 |....4V|
00000014
root@NBG7815:~# hexdump -s 20 -n 6 -C /lib/firmware/ath11k/IPQ8074/hw2.0/cal-ahb-c000000.wifi.bin
00000014  00 03 7f 12 34 57                                 |....4W|
0000001a
root@NBG7815:~# hexdump -s 26 -n 6 -C /lib/firmware/ath11k/IPQ8074/hw2.0/cal-ahb-c000000.wifi.bin
0000001a  00 03 7f 12 34 58                                 |....4X|
00000020
```

The last 4 digits of each are alternated on each reboot. While I do not understand why we would need label_mac here. Because label MAC is ZyXEL DEV MAC (starting with BC:CF:4F:xx:xx:xx, attached on WAN/LAN). While 00 03 7f is Atheros DEV MAC.

testuser7 replied:
MAC ids used in function starts from `1` not `0`. That's why third MAC is not set.

pwned-pixel replied:
Y, that what I thought already. I've tried several combinations. But all the same. Then I realized this file is meant to not change across reboots. Oh dear.
Made a "normal" build now with all my settings and applications. Everything is working now. Thank you very much for your endurance in guiding me.

Final settings I've use are:

```
	zyxel,nbg7815)
		caldata_extract "0:art" 0x1000 0x20000
		label_mac=$(get_mac_label)
		ath11k_patch_mac $(macaddr_add $label_mac 2) 1 
		ath11k_patch_mac $(macaddr_add $label_mac 3) 2 
		ath11k_patch_mac $(macaddr_add $label_mac 4) 3 
		ath11k_fix_caldata
		;;
```

I have one question left. As I've made a pull request for the MAC fix with the "ieee80211/11_fix_wifi_mac" (https://github.com/openwrt/openwrt/pull/15025) approach. Should I close it or keep it open and modify it accordingly after your patch gets accepted? What is the advise/usual procedure?

testuser7 replied:
As the interfaces are numbered from 0, I changed the parameters for function from 0 to 5.

testuser7 commented:
@robimarko What else needs to be done to merge these changes?
Even in qsdk they correct the `nvMacFlag` flag: https://github.com/FanFansfan/qsdk-5.4/blob/master/qca/feeds/wlan_utils/wifi-scripts/files/setmac#L122

robimarko commented:
I trust you will follow-up with fixes if needed :)

testuser7 commented:
Sure, thanks!
